File: Ravi/Microsoft/Day 3/CourseSchedule.java
class Solution {
    
    boolean dfs(int node, int[] vis, int[] dfsvis, ArrayList<Integer> adj[]) {
        vis[node] = 1;
        dfsvis[node] = 1;
        for(int i : adj[node]) {
            if(vis[i] == 0) 
            {
                if(dfs(i, vis, dfsvis, adj)) return true;
            } else if(dfsvis[i] == 1) return true;
        }
        dfsvis[node] = 0;
        return false;
    }
    
    public boolean canFinish(int n, int[][] p) {
        
        ArrayList<Integer> adj[] = new ArrayList[n];
        
        for(int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        
        int []vis = new int[n];
        int []dfsvis = new int[n];
        
        for(int i = 0; i < p.length; i++) {
            adj[p[i][0]].add(p[i][1]);
        }
        
        for(int i = 0; i < n; i++) {
            if(vis[i] == 0) {
                if(dfs(i, vis, dfsvis, adj)) return false;
            }
        }
        
        return true;
        
    }
}

File: Sudesh_R/Microsoft/Day3/course-schedule.cpp
class Solution {
public:
vector<int>adj[100005];
bool vis[100005];
bool dvis[100005];
    bool solve(int node )
    {
        vis[node] = 1;
        dvis[node] = 1;
        for(auto it : adj[node])
        {
            if(!vis[it])
            {
                if(solve(it))
                {
                    return true;
                }
            }
            else if(dvis[it])
            {
                return true;
            }

        }
        dvis[node] = 0;
        return false;
        
    }
    bool canFinish(int n, vector<vector<int>>& p) 
    {
        if(p.size()==0)return 1;
        for(int i  = 0;i<p.size();i++)
        {
            adj[p[i][0]].push_back(p[i][1]);
        }
      bool  x = 0;
        for(int i = 0 ;i<n;i++)
        {
            if(!vis[i])
            {
               if(solve(i )) return 0 ;
            }
        }
        return 1;
    }
};
